app/src/main/java/org/kiwix/kiwixmobile/localFileTransfer/ReceiverDevice.kt

```diff
@@ -0,0 +1,109 @@
+/*
+ * Kiwix Android
+ * Copyright (c) 2019 Kiwix <android.kiwix.org>
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+package org.kiwix.kiwixmobile.localFileTransfer
+
+import android.util.Log
+import android.widget.Toast
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.isActive
+import kotlinx.coroutines.withContext
+import org.kiwix.kiwixmobile.core.BuildConfig
+import org.kiwix.kiwixmobile.core.R
+import org.kiwix.kiwixmobile.core.extensions.toast
+import org.kiwix.kiwixmobile.localFileTransfer.WifiDirectManager.Companion.copyToOutputStream
+import java.io.File
+import java.io.FileOutputStream
+import java.io.IOException
+import java.net.ServerSocket
+
+/**
+ * Helper class for the local file sharing module.
+ *
+ * Once the handshake has successfully taken place, this async-task is used to receive files from
+ * the sender device on the FILE_TRANSFER_PORT port. No. of files to be received (and their names)
+ * are learnt beforehand during the handshake.
+ *
+ * A single Task is used for the entire file transfer (the server socket accepts connections as
+ * many times as the no. of files).
+ */
+internal class ReceiverDevice(private val wifiDirectManager: WifiDirectManager) {
+  private var incomingFileName: String? = null
+  suspend fun receive() = withContext(Dispatchers.IO) {
+    try {
+      ServerSocket(WifiDirectManager.FILE_TRANSFER_PORT).use { serverSocket ->
+        Log.d(TAG, "Server: Socket opened at " + WifiDirectManager.FILE_TRANSFER_PORT)
+        val zimStorageRootPath = wifiDirectManager.zimStorageRootPath
+        val fileItems = wifiDirectManager.getFilesForTransfer()
+        var isTransferErrorFree = true
+        if (BuildConfig.DEBUG) Log.d(TAG, "Expecting " + fileItems.size + " files")
+        var fileItemIndex = 0
+        while (fileItemIndex < fileItems.size && this.isActive) {
```

Comment:
what was the bug?
```
if (!dirs.exists() && !dirs.mkdirs()) {
                Log.d(TAG, "ERROR: Required parent directories couldn't be created")
                isTransferErrorFree = false
                fileItemIndex++
              }
```
that code is a bit ridiculous because if that happens then I am pretty sure one of the following lines would throw an IO exception which means
```
catch (e: IOException) {
            Log.e(TAG, e.message)
            isTransferErrorFree = false
            publishProgress(fileItemIndex, FileItem.FileStatus.ERROR)
          }
```
we would set the next item index as error. 

If I am reading it correctly

Reply:
@macgills I don't know if I change it doesn't show the new file list.

Reply:
We don't let bugs beat us (okay maybe sometimes), you set a breakpoint in forEach and what happened? What did the logs say?

This is what receives the data of the files, if your receiver device is not displaying the list of files it is about to receive then that is an issue in the handshake where it sends the filenames over in the metadata exchange

Reply:
@macgills I just tried to manually debug it. i looked over logs and I found this
<details>
<summary>logs</summary>

```
le D/WifiDirectManager: WiFi P2P state changed - false
27643-27643/org.kiwix.kiwixmobile D/WifiDirectManager: WiFi P2P state changed - true
27643-27643/org.kiwix.kiwixmobile D/WifiDirectManager: Peer Status: 3
27643-27643/org.kiwix.kiwixmobile D/WifiDirectManager: Peer Status: 3
27643-27643/org.kiwix.kiwixmobile D/WifiDirectManager: P2P peers changed
27643-27643/org.kiwix.kiwixmobile D/WifiDirectManager: P2P peers changed
27643-27643/org.kiwix.kiwixmobile D/WifiDirectManager: Peer Status: 0
27643-27643/org.kiwix.kiwixmobile D/WifiDirectManager: Starting handshake
27643-27643/org.kiwix.kiwixmobile D/WifiDirectManager: ReceiverDeviceAsyncTask complete
27643-27643/org.kiwix.kiwixmobile D/WifiDirectManager: Disconnect successful
```
</details>

Reply:
```
 if (BuildConfig.DEBUG) Log.d(TAG, "Expecting " + fileItems.size + " files")
```
where is this line? there should be far more logs

Reply:
you are setting breakpoints and attaching a debugger?

Reply:
> you are setting breakpoints and attaching a debugger?

No, I just looked at the code and filter the logs. Since this will take more time and today is Friday 😆

Reply:
> you are setting breakpoints and attaching a debugger?

No, I just looked at the code and filter the logs

Reply:
I found the bug.When i refactored the code I accidently forget about this 
```java
incomingFileName = fileItem.fileName
```
so I stored `FileItem` as fileName that's why the File was sent completely but it didn't create `File` object in receiver device.

Reply:
[done](https://github.com/kiwix/kiwix-android/pull/2389/commits/a33558dd2c4d13e1afcde3136718718ca69f4f81)